api/utils/route.go

```diff
@@ -0,0 +1,155 @@
+	return func(opts *sshRouteOptions) {
+		opts.caseInsensitive = cir
+	}
+}
+
+// SSHRouteMatcher is a helper used to decide if an ssh dial request should match
+// a given server. This is broken out of proxy.Router as a standalone helper in order
+// to let other parts of teleport easily find matching servers when generating
+// error messages or building access requests.
+type SSHRouteMatcher struct {
+	sshRouteOptions
+	targetHost     string
+	targetPort     string
+	ips            []string
+	matchServerIDs bool
+}
+
+// NewSSHRouteMatcher builds a new matcher for ssh routing decisions.
+func NewSSHRouteMatcher(host, port string, opts ...SSHRouteOption) SSHRouteMatcher {
+	var options sshRouteOptions
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	_, err := uuid.Parse(host)
+	dialByID := err == nil || IsEC2NodeID(host)
+
+	ips, _ := net.LookupHost(host)
+
+	return SSHRouteMatcher{
+		sshRouteOptions: options,
+		targetHost:      host,
+		targetPort:      port,
+		ips:             ips,
+		matchServerIDs:  dialByID,
+	}
+}
+
+// sshServer is an interface describing the subset of the types.Server interface
+// required to make a routing decision.
+type sshServer interface {
+	GetName() string
+	GetHostname() string
+	GetAddr() string
+	GetUseTunnel() bool
+}
+
+func (m *SSHRouteMatcher) RouteToServer(server sshServer) bool {
+	// if host is a UUID or EC2 ID match only
+	// by server name and treat matches as unambiguous
+	if m.matchServerIDs && server.GetName() == m.targetHost {
+		return true
+	}
+
+	// if the server has connected over a reverse tunnel
+	// then match only by hostname.
+	if server.GetUseTunnel() {
+		return m.routeToHostname(server.GetHostname())
+	}
+
+	ip, nodePort, err := net.SplitHostPort(server.GetAddr())
+	if err != nil {
+		return false
+	}
+
+	if (m.targetHost == ip || m.routeToHostname(server.GetHostname()) || slices.Contains(m.ips, ip)) &&
+		(m.targetPort == "" || m.targetPort == "0" || m.targetPort == nodePort) {
+		return true
+	}
+
+	return false
+}
+
+// routeToHostname helps us perform a special kind of case-insensitive comparison. SSH certs do not generally
+// treat principals/hostnames in a case-insensitive manner. This is often worked-around by forcing all principals and
+// hostnames to be lowercase. For backwards-compatibility reasons, teleport must support case-sensitive routing by default
+// and can't do this. Instead, teleport nodes whose hostnames contain uppercase letters will present certs that include both
+// the literal hostname and a lowered version of the hostname, meaning that it is sane to route a request for host 'foo' to
+// host 'Foo', but it is not sane to route a request for host 'Bar' to host 'bar'.
+func (m *SSHRouteMatcher) routeToHostname(principal string) bool {
+	if !m.caseInsensitive {
+		return m.targetHost == principal
+	}
+
+	if len(m.targetHost) != len(principal) {
+		return false
+	}
+
+	// the below is modeled off of the fast ASCII path of strings.EqualFold
```

Comment:
What's wrong with just using `stringsEqualFold` here?

Reply:
Couple reasons:

1. Hostname insensitivity, where it's used, is typically defined as only insensitivity to ASCII casing.  `strings.EqualFold` uses unicode case-folding which is insensitive across a much broader range of possible input characters.  Go's own TLS libraries are only case-insensitive for ASCII casing.
2. The routing logic in the PR uses a modified version of insensitivity that is insensitive only to uppercase characters in the server principal, but not in the target host arg.  We don't route `tsh ssh Foo` to host `foo`, but we do route `tsh ssh foo` to host `Foo`.  This is because we aren't actually modifying how SSH certificate validation works in order to implement this feature (which *is* case sensitive in both `openssh` and go's `crypto/ssh`, and while I couldn't find a conclusive answer as to wether case-sensitivity is a requirement, I also couldn't find any RFCs that mention permitting case-insensitivity of any kind in SSH certificate validation).  Instead, after this PR, all teleport hosts whose hostnames contain uppercase ASCII characters will include a lowered-ascii version of the hostname in their principals as well (i.e. a host with hostname `Foo` will allocate a cert with principals `Foo` _and_ `foo`).  We don't do the opposite (adding uppercase principals to the certs of an agent with a lowercase hostname), because in order to do that correctly we'd need to permute all possible combinations of upper and lowercase ascii characters for that hostname.